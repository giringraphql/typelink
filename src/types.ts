export type ResolvedValue<T> = T | Promise<T>;
export type ResolvedList<T> = T[] | Promise<T[]> | Promise<T>[];

export interface ConcreteClass<T = any> {
  new(...args: any[]): T;
}

export type Instantiator<TClass = any> = (value: { [key: string]: any }) => TClass;

export function defaultInputFieldInstantiator(value: any) {
  return value;
}

export type Thunk<T> = () => T;
