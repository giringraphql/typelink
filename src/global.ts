import { loadFallbackRootTypes } from './definition/rootTypes';
import { loadBuiltInScalar } from './definition/scalar';
import { DefinitionKind } from './metadata/Definition';
import { MetadataStorage } from './metadata/MetadataStorage';
import { DefinitionParser } from './sdl/ast';
import { coerceType } from './type-expression/coerceType';
import { TypeExpression } from './type-expression/TypeExpression';
import { Thunk } from './types';
import { TypeArg } from './type-expression/types';


/**
 * Create a new MetadataStorage initialized with default metadata
 */
export function createMetadataStorage() {
  const storage = new MetadataStorage();
  loadBuiltInScalar(storage);
  loadFallbackRootTypes(storage);
  return storage;
}

/**
 * Global MetadataStorage used by default.
 */
export let globalMetadataStorage: MetadataStorage;

export function getGlobalMetadataStorage() {
  if (!globalMetadataStorage) {
    globalMetadataStorage = createMetadataStorage();
  }
  return globalMetadataStorage;
}

/**
 * Get a GraphQLType instance from the given storage or default
 * global metadata storage.
 * @param typeArg
 * @param storage
 */
export function getType(typeArg: TypeExpression | TypeArg, kind: DefinitionKind = 'any'): any {
  const storage = getGlobalMetadataStorage();
  return coerceType(typeArg).getType({ kind, storage });
}

/**
 * Define a type linked to decorated class and add it to the given
 * storage or default global metadata storage.
 * @param metadataOrThunk
 */
export function defineType(parsersOrThunk: DefinitionParser[] | Thunk<DefinitionParser[]>) {
  const storage = getGlobalMetadataStorage();
  return function defDecoratorFn<T extends Function>(definitionClass: T) {
    if (Array.isArray(parsersOrThunk)) {
      parsersOrThunk.forEach(parser => {
        storage.registerMetadata(parser.parse(definitionClass));
      });
    } else {
      storage.deferRegister(() => {
        parsersOrThunk().forEach(parser => {
          storage.registerMetadata(parser.parse(definitionClass));
        });
      });
    }
    return definitionClass;
  };
}
