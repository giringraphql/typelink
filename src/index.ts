export { gql } from './sdl/gql';

export * from './definition/GraphQLTypeIndex';
export * from './definition/InputType';
export * from './definition/InterfaceType';
export * from './definition/ObjectType';
export * from './definition/rootTypes';
export * from './definition/scalar';
export * from './definition/SubscriptionType';


export * from './reference/Field';
export * from './reference/Implement';
export * from './reference/InputField';
export * from './global';

export * from './type-expression/DefinitionTypeExpression';
export * from './type-expression/GraphQLTypeExpression';
export * from './type-expression/TypeExpression';
export * from './type-expression/structure';
export * from './type-expression/coerceType';
export * from './type-expression/types';

export * from './metadata/Definition';
export * from './metadata/MetadataStorage';
export * from './metadata/Reference';

export * from './relay/connection';
export * from './relay/mutation';
export * from './relay/node';
