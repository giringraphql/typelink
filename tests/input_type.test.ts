import { GraphQLSchema, graphql, printSchema } from 'graphql';

import { getType, gql, defineType } from '../src';


@defineType(gql`
  input NameInput {
    firstName: String!
    lastName: String!
  }
  type Name {
    firstName: String!
    lastName: String!
  }
`)
class Name {

  firstName: string;
  lastName: string;

  get fullName() {
    return `${this.firstName} ${this.lastName}`;
  }
}

let instantiationCount = 0;

@defineType(gql`
  input PersonInput {
    address: String!
    name: ${Name}
  }
  type Person {
    address: String!
    name: ${Name}
  }
`)
class Person {
  address: string;
  name: Name;
  constructor() {
    instantiationCount ++;
  }
}

@defineType(gql`
  type Group {
    echoPerson(person: ${Person}): ${Person}!
  }
`)
class Group {
  echoPerson(args: { person: Person }) {
    return args.person;
  }
}

@defineType(gql`
  type Query {            # resolved to NameInput
    formatFullName(input: ${Name}): String!
                                    # resolved to PersonInput
    personInputInstantiated(person: ${Person}): Boolean!
    personNonNullInputWorks(person: ${Person}!): Boolean!
    personNonNullListInputWorks(people: [${Person}!]): Boolean!

    echoPerson(person: ${Person}): ${Person}! # resolved to Person
    tenGroups: [${Group}]!
  }
`)
class Query {
  static formatFullName(_source: null, args: { input: Name }) {
    return args.input.fullName;
  }
  static personInputInstantiated(_source: null, args: { person: Person }) {
    return (args.person instanceof Person) && (args.person.name instanceof Name);
  }
  static personNonNullInputWorks(_source: null, args: { person: Person }) {
    return (args.person instanceof Person) && (args.person.name instanceof Name);
  }
  static personNonNullListInputWorks(_source: null, args: { people: Person[] }) {
    return args.people.reduce((res, person) => {
      return res && (person instanceof Person) && (person.name instanceof Name);
    }, true);
  }
  static echoPerson(_source: null, args: { person: Person }) {
    return args.person;
  }
  static tenGroups() {
    const groups = [];
    for (let i = 0; i < 10; i++) {
      groups.push(new Group());
    }
    return groups;
  }
}

const schema = new GraphQLSchema({
  query: getType(Query),
});

describe('Input type', () => {
  it('generates schema and works as expected', async () => {

    expect(printSchema(schema)).toMatchSnapshot();

    const results = await graphql({ schema, source: `
      query {
        formatFullName(input: { firstName: "Foo", lastName: "Bar" })
        personInputInstantiated(person: {
          address: "A",
          name: {
            firstName: "Foo",
            lastName: "Bar"
          }
        })
        personNonNullInputWorks(person: {
          address: "A",
          name: {
            firstName: "Foo",
            lastName: "Bar"
          }
        })
        personNonNullListInputWorks(people: [
          {
            address: "A",
            name: {
              firstName: "Foo",
              lastName: "Bar"
            }
          },
          {
            address: "B",
            name: {
              firstName: "Foo",
              lastName: "Bar"
            }
          }
        ])

        echoPerson(person: {
          address: "A",
          name: {
            firstName: "Foo",
            lastName: "Bar"
          }
        }) {
          address
          name {
            firstName
            lastName
          }
        }
      }
    ` });

    expect(results).toEqual({ data: {
      formatFullName: 'Foo Bar',
      personInputInstantiated: true,
      personNonNullInputWorks: true,
      personNonNullListInputWorks: true,
      echoPerson: {
        address: 'A',
        name: {
          firstName: 'Foo',
          lastName: 'Bar'
        }
      }
    } });
  });

  it('should not instantiate input type which is already cached', async () => {
    instantiationCount = 0;
    await graphql({ schema, source: `
      query($person: PersonInput) {
        tenGroups {
          echoPerson(person: $person) {
            address
          }
        }
      }
    `, variableValues: {
      person: {
        address: 'A',
        name: {
          firstName: 'Foo',
          lastName: 'Bar'
        }
      },
    }}); // giving argument as variable lets us avoid extra instantiations
    expect(instantiationCount).toBe(1);
  });
});
