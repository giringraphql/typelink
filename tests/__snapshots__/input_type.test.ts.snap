// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input type generates schema and works as expected 1`] = `
"type Group {
  echoPerson(person: PersonInput): Person!
}

type Name {
  firstName: String!
  lastName: String!
}

input NameInput {
  firstName: String!
  lastName: String!
}

type Person {
  address: String!
  name: Name
}

input PersonInput {
  address: String!
  name: NameInput
}

type Query {
  formatFullName(input: NameInput): String!
  personInputInstantiated(person: PersonInput): Boolean!
  personNonNullInputWorks(person: PersonInput!): Boolean!
  personNonNullListInputWorks(people: [PersonInput!]): Boolean!
  echoPerson(person: PersonInput): Person!
  tenGroups: [Group]!
}
"
`;
